lib/quantization/src/quantile.rs

```diff
@@ -61,11 +64,140 @@ pub(crate) fn find_quantile_interval<'a>(
+    }
+
+    let (_, selected_vectors) = take_random_vectors(vector_data, count, P2_SAMPLE_SIZE);
+
+    for vector in selected_vectors {
+        if stopped.load(Ordering::Relaxed) {
+            return Err(EncodingError::Stopped);
+        }
+
+        let vector = vector.as_ref();
+        debug_assert_eq!(
+            vector.len(),
+            dim,
+            "Vector length does not match the expected dimension"
+        );
+
+        for (i, min_max) in result.iter_mut().enumerate() {
+            if vector[i] < min_max.0 {
+                min_max.0 = vector[i];
+            }
+            if vector[i] > min_max.1 {
+                min_max.1 = vector[i];
+            }
+        }
+    }
+
+    for min_max in result.iter_mut() {
+        if min_max.0 == f32::MAX || min_max.1 == f32::MIN {
+            *min_max = (0.0, 0.0);
+        }
+    }
+
+    Ok(result)
+}
+
+fn find_interval_per_coordinate_single_thread<'a>(
+    vector_data: impl Iterator<Item = impl AsRef<[f32]> + 'a> + Clone,
+    dim: usize,
+    quantile: f32,
+    stopped: &AtomicBool,
+) -> Result<Vec<(f32, f32)>, EncodingError> {
+    let mut p_squares: Vec<(P2Quantile<P2_MARKERS>, P2Quantile<P2_MARKERS>)> = (0..dim)
+        .map(|_| {
+            (
+                P2Quantile::<P2_MARKERS>::new(f64::from(quantile)).unwrap(),
+                P2Quantile::<P2_MARKERS>::new(1.0 - f64::from(quantile)).unwrap(),
+            )
+        })
+        .collect();
+
+    let (_, selected_vectors) = take_random_vectors(vector_data, usize::MAX, P2_SAMPLE_SIZE);
+
+    for vector in selected_vectors {
+        if stopped.load(Ordering::Relaxed) {
+            return Err(EncodingError::Stopped);
+        }
+
+        let vector = vector.as_ref();
+        debug_assert_eq!(
+            vector.len(),
+            dim,
+            "Vector length does not match the expected dimension"
+        );
+
+        for d in 0..dim {
+            p_squares[d].0.push(f64::from(vector[d]));
+            p_squares[d].1.push(f64::from(vector[d]));
+        }
+    }
+    let mut result = Vec::with_capacity(dim);
+    for (p_square_min, p_square_max) in p_squares {
+        result.push((
+            p_square_min.estimate() as f32,
+            p_square_max.estimate() as f32,
+        ));
+    }
+    Ok(result)
+}
+
+fn take_random_vectors<'a>(
+    vector_data: impl Iterator<Item = impl AsRef<[f32]> + 'a>,
+    count: usize,
+    sample_size: usize,
+) -> (usize, impl Iterator<Item = impl AsRef<[f32]> + 'a>) {
+    let slice_size = std::cmp::min(count, sample_size);
+    let permutor = Permutor::new(count as u64);
+    let mut selected_vectors: Vec<usize> = permutor.map(|i| i as usize).take(slice_size).collect();
+    selected_vectors.sort_unstable();
+
+    (
+        slice_size,
+        vector_data
+            .enumerate()
+            .filter_map(move |(vector_index, vector)| {
+                if selected_vectors.binary_search(&vector_index).is_ok() {
```

Comment:
It could be implemented as linear operation because both `selected_vectors`'s values and `vector_index` is increasing.